src/lib/mdns/ServicePool.cpp

```diff
@@ -0,0 +1,272 @@
+{
+    Clear();
+    mService               = service;
+    mNodeId                = nodeId;
+    mFabricId              = fabricId;
+    service.mTextEntryies  = nullptr;
+    service.mTextEntrySize = 0;
+    mHasValue              = true;
+    mHasPending            = false;
+
+    return CHIP_NO_ERROR;
+}
+
+void ServicePool::Entry::Clear()
+{
+    if (mHasValue)
+    {
+        FreeTextEntries(&mService);
+    }
+    mHasValue   = false;
+    mHasPending = false;
+}
+
+ServicePool::Entry::~Entry()
+{
+    Clear();
+}
+
+ServicePool::Entry & ServicePool::FindAvailableSlot(size_t hashValue)
+{
+    size_t i   = hashValue;
+    bool found = false;
+
+    do
+    {
+        if (!mEntries[i].mHasValue)
+        {
+            found = true;
+            break;
+        }
+
+        i++;
+        i %= kServicePoolCapacity;
+    } while (i != hashValue);
+
+    if (!found)
+    {
+        i = hashValue;
+    }
+
+    return mEntries[i];
+}
+
+CHIP_ERROR ServicePool::AddService(uint64_t nodeId, uint64_t fabricId, const MdnsService & service)
+{
+    return FindAvailableSlot(nodeId % kServicePoolCapacity).Emplace(service, nodeId, fabricId);
+}
+
+CHIP_ERROR ServicePool::AddService(uint64_t nodeId, uint64_t fabricId, MdnsService && service)
+{
+
+    return FindAvailableSlot(nodeId % kServicePoolCapacity).Emplace(std::move(service), nodeId, fabricId);
+}
+
+CHIP_ERROR ServicePool::RemoveService(uint64_t nodeId, uint64_t fabricId)
+{
+    Entry * entry = FindService(nodeId, fabricId);
+
+    if (!entry)
+    {
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+    entry->Clear();
+    // Because of hash map open addressing, we still need to mark it as "Has pending" to make the search continue
+    entry->mHasPending = true;
+    mLazyDeleteCount++;
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR ServicePool::UpdateService(uint64_t nodeId, uint64_t fabricId, const MdnsService & service)
+{
+    Entry * entry = FindService(nodeId, fabricId);
+
+    if (!entry)
+    {
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    return entry->Emplace(service, nodeId, fabricId);
+}
+
+ServicePool::Entry * ServicePool::FindService(uint64_t nodeId, uint64_t fabricId)
+{
+    size_t hashValue = nodeId % kServicePoolCapacity;
+    size_t i         = hashValue;
+
+    do
+    {
+        if (!mEntries[i].mHasValue && !mEntries[i].mHasPending)
+        {
+            break;
+        }
+        if (mEntries[i].mHasValue && mEntries[i].mNodeId == nodeId && mEntries[i].mFabricId == fabricId)
+        {
+            return &mEntries[i];
+        }
+        i++;
+        i %= kServicePoolCapacity;
+    } while (i != hashValue);
+
+    return nullptr;
+}
+
+void ServicePool::Clear()
+{
+    for (Entry & entry : mEntries)
+    {
+        entry.Clear();
+    }
+    mLazyDeleteCount = 0;
+}
+
+void ServicePool::ReHash()
+{
+    Entry copyEntries[kServicePoolCapacity];
+
+    for (size_t i = 0; i < kServicePoolCapacity; i++)
```

Comment:
One thought worth testing for codesize impact, though I expect it's no worse than this: make `mEntries` and this temporary both `std::Array<Entry, kServicePoolCapacity>`.  You'll probably want to typedef that (as `EntryList` or something?).  Then you can do:
```
  EntryList copyEntries(std::move(mEntries));
```
and that will invoke the move constructors on all the entries as needed, etc.  Then you don't need the fragile move assignment operator.